libraries/AP_Common/Semaphore.h

```diff
@@ -28,20 +26,25 @@
 
   The WITH_SEMAPHORE() macro can be used with either type of semaphore
  */
+
+namespace AP_HAL {
+class Semaphore;
+}
+
 class WithSemaphore {
 public:
-    WithSemaphore(HAL_Semaphore &mtx) :
-    _mtx(mtx)
-    {
+    WithSemaphore(AP_HAL::Semaphore *mtx) : WithSemaphore(*mtx) { }
+
+    WithSemaphore(AP_HAL::Semaphore &mtx) : _mtx(mtx) {
         _mtx.take_blocking();
     }
 
     ~WithSemaphore() {
         _mtx.give();
     }
 private:
-    HAL_Semaphore &_mtx;
```

Comment:
I agree.

Reply:
But.... what if we made it like this:

every HAL does:

```c++
namespace AP_HAL {
using Semaphore = PX4::Semaphore
}
```

It would be even better if we could rename it to Mutex, since every HAL is actually using a mutex, not a semaphore.